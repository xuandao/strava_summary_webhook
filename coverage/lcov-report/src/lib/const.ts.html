
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for src/lib/const.ts</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <link rel="shortcut icon" type="image/x-icon" href="../../favicon.png" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
    
<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="../../index.html">All files</a> / <a href="index.html">src/lib</a> const.ts</h1>
        <div class='clearfix'>
            
            <div class='fl pad1y space-right2'>
                <span class="strong">100% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>12/12</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">100% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>0/0</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">100% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>0/0</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">100% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>12/12</span>
            </div>
        
            
        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
        <template id="filterTemplate">
            <div class="quiet">
                Filter:
                <input oninput="onInput()" type="search" id="fileSearch">
            </div>
        </template>
    </div>
    <div class='status-line high'></div>
    <pre><table class="coverage">
<tr><td class="line-count quiet"><a name='L1'></a><a href='#L1'>1</a>
<a name='L2'></a><a href='#L2'>2</a>
<a name='L3'></a><a href='#L3'>3</a>
<a name='L4'></a><a href='#L4'>4</a>
<a name='L5'></a><a href='#L5'>5</a>
<a name='L6'></a><a href='#L6'>6</a>
<a name='L7'></a><a href='#L7'>7</a>
<a name='L8'></a><a href='#L8'>8</a>
<a name='L9'></a><a href='#L9'>9</a>
<a name='L10'></a><a href='#L10'>10</a>
<a name='L11'></a><a href='#L11'>11</a>
<a name='L12'></a><a href='#L12'>12</a></td><td class="line-coverage quiet"><span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span></td><td class="text"><pre class="prettyprint lang-js">export const GIT_OWNER = 'xuandao';
export const GIT_REPO = 'running_page';
export const GIT_ACTION_TYPE = '"webhook-event';
export const GIT_PRIVATE_ACCESS_TOKEN= 'Bearer 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';
&nbsp;
export const STRAVA_CLIENT_ID = "75603";
export const STRAVA_CLIENT_SECRET = "c21644951d5f40965f1767658d6197186d47346e";
export const STRAVA_REFRESH_TOKEN = "dbf00fd1fd6a47841b705a67f43d8d9e073e0c46";
&nbsp;
export const TYPE_OUTDOOR = "OutDoor";
export const TYPE_INDOOR = "InDoor";
export const TYPE_ALL = "All";</pre></td></tr></table></pre>

                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank" rel="noopener noreferrer">istanbul</a>
                at 2024-02-14T02:36:55.905Z
            </div>
        <script src="../../prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="../../sorter.js"></script>
        <script src="../../block-navigation.js"></script>
    </body>
</html>
    