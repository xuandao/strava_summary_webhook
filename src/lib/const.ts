const ENV = process.env;
export const GIT_OWNER = ENV.GIT_OWNER || "";
export const GIT_REPO = ENV.GIT_REPO || "";
export const GIT_ACTION_TYPE = ENV.GIT_ACTION_TYPE || "";
export const GIT_PRIVATE_ACCESS_TOKEN= ENV.GIT_PRIVATE_ACCESS_TOKEN || "";

export const STRAVA_CLIENT_ID = ENV.STRAVA_CLIENT_ID || "";
export const STRAVA_CLIENT_SECRET = ENV.STRAVA_CLIENT_SECRET || "";
export const STRAVA_REFRESH_TOKEN = ENV.STRAVA_REFRESH_TOKEN || "";

export const TYPE_OUTDOOR = "OutDoor";
export const TYPE_INDOOR = "InDoor";
export const TYPE_ALL = "All";