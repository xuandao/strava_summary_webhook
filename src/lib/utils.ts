import type { NextApiRequest } from 'next'

export function getCurrentTimestampMillis(): number {
    return Date.now();
}

export function safeJSON(str:string){
    try{
        JSON.parse(str)
    } catch(e) {
        // ingore
    }
}

export function requestArgsLog(req: NextApiRequest){
    let args;
    switch(req.method){
        case 'GET':
            args = req.query;
            break;
        case 'POST':
            args = req.body;
            break;
    }

    console.log(`requestArguments: ${JSON.stringify(args)}`);
}