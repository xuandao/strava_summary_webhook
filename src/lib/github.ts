import { Octokit } from "@octokit/core";
import {GIT_OWNER, GIT_REPO, GIT_ACTION_TYPE, GIT_PRIVATE_ACCESS_TOKEN} from "@/lib/const";
const octokit = new Octokit({
    auth: GIT_PRIVATE_ACCESS_TOKEN
})

/**
 * trigger github actions
 * @returns 
 */
export async function triggerDataSyncAction() {
    await octokit.request('POST /repos/{owner}/{repo}/dispatches', {
        owner: GIT_OWNER,
        repo: GIT_REPO,
        event_type: GIT_ACTION_TYPE,
        client_payload: {},
        headers: {
            'X-GitHub-Api-Version': '2022-11-28'
        }
    })
}