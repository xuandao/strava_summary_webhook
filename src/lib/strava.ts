import { Strava, DetailedActivity, SummaryActivity } from "strava";
import {STRAVA_CLIENT_ID, STRAVA_CLIENT_SECRET, STRAVA_REFRESH_TOKEN, TYPE_ALL, TYPE_INDOOR, TYPE_OUTDOOR} from "@/lib/const";
import { getCurrentTimestampMillis } from "@/lib/utils";

const STRAVA_CLIENT = new Strava({
  client_id: STRAVA_CLIENT_ID,
  client_secret: STRAVA_CLIENT_SECRET,
  refresh_token: STRAVA_REFRESH_TOKEN
});

/**
 * 获取单个活动信息
 * @param id 
 * @returns 
 */
export async function getActivityById(id:number): Promise<DetailedActivity> {
  return await STRAVA_CLIENT.activities.getActivityById({id})
}

/**
 * 获取时间范围的所有活动列表
 * @param type 
 * @param days 
 * @returns 
 */
export async function getLoggedInAthleteActivities(type: string, days: number): Promise<SummaryActivity[]>{
  // 4 weeks, 28 days
  const DATE_GAP = 1000 * 60 * 60 * 24 * days;
  const after = (getCurrentTimestampMillis() - DATE_GAP) / 1000;

  const ret:SummaryActivity[] = await STRAVA_CLIENT.activities.getLoggedInAthleteActivities({
      after,
      per_page: 50
  });

  // 过滤所需类型
  return ret.filter(item => item.type === type)
}

/**
 * 更新活动的 description
 * @param id 
 * @param description 
 * @returns 
 */
export async function updateActivityDescription(id:number, description:string): Promise<void> {
    await STRAVA_CLIENT.activities.updateActivityById({
      id: id,
      description
    });
}

/**
 * 订阅 webhook
 * @param callback_url 
 */
export async function createSubscription(callback_url: string): Promise<any>{
  return await STRAVA_CLIENT.subscriptions.createSubscription({
    callback_url,
    verify_token: "STRAVA"
  })
}

/**
 * 更新活动 summary
 * @returns 
 */
export function generateActivitiesSummary(curActivity: DetailedActivity, allActivitiesList: SummaryActivity[]): string{
  const subType = curActivity.map.summary_polyline !== undefined ? TYPE_OUTDOOR : TYPE_INDOOR;

    // 当前类型的 Summary
    const subActivitesList = getSubtypeActivites(subType, allActivitiesList);
    const sameTypeSummary: string = rendersummaryResult(
      subType,
      curActivity,
      subActivitesList
    );

    // 所有类型活动的 Summary
    const allSummary: string = rendersummaryResult(
      TYPE_ALL,
      curActivity,
      allActivitiesList
    );

    const description = `${sameTypeSummary}\n\n${allSummary}`;

    return description;
}

/**
 * 获取子类型的活动列表
 * @param subType 
 */
function getSubtypeActivites(subType:string, allActivitiesList: SummaryActivity[]): SummaryActivity[] {
  if(subType === TYPE_OUTDOOR){
    return allActivitiesList.filter(item => !!item.map.summary_polyline);
  }

  return allActivitiesList.filter(item => !item.map.summary_polyline);
}

/**
 * 渲染 Summary
 * @param type 
 * @param curActivity 
 * @param allActivitiesList 
 * @returns 
 */
function rendersummaryResult(curType: string, curActivity: DetailedActivity, allActivitiesList: SummaryActivity[]) {
  const DATA_LEN = allActivitiesList.length;

  //总里程
  const TOTAL_DISTANCE = allActivitiesList.reduce(
    (curValue, item) => curValue + item.distance,
    0
  );
  //总运动时间
  const TOTAL_MTime = allActivitiesList.reduce(
    (curValue, item) => curValue + item.moving_time,
    0
  );
  //里程
  const AVG_DISTANCE =
  allActivitiesList.reduce((curValue, item) => curValue + item.distance, 0) /
    DATA_LEN;
  // 心率
  const AVG_HR =
  allActivitiesList.reduce(
      (curValue, item) => curValue + item.average_heartrate,
      0
    ) / DATA_LEN;
  // 步频
  const AVG_CADENCE =
  allActivitiesList.reduce((curValue, item) => curValue + item.average_cadence, 0) /
    DATA_LEN;
  // 运动时间
  const AVG_MTime =
  allActivitiesList.reduce((curValue, item) => curValue + item.moving_time, 0) /
    DATA_LEN;
  //负荷
  const AVG_SCORE =
  allActivitiesList.reduce((curValue, item) => curValue + item.suffer_score, 0) /
    DATA_LEN;

  const EMOJI_UNDONE = "❌";
  const EMOJI_DONE = "✅";
  const EMOJI_UP = "";
  const EMOJI_DOWN = "⤵️";

  return `
${curType} SUMMARY${curType === TYPE_ALL ? "(All Runs & IndoorRuns)" : ""}:
    Completed: ${DATA_LEN} runs
    Total Distance:: ${(TOTAL_DISTANCE / 1000).toFixed(2)}km (Avg.: ${(
    AVG_DISTANCE / 1000
  ).toFixed(2)}km ${
    curActivity.distance > AVG_DISTANCE ? EMOJI_DONE : EMOJI_UNDONE
  })
    Total Moving Time: ${(TOTAL_MTime / 60).toFixed(2)}Mins (Avg.: ${(
    AVG_MTime / 60
  ).toFixed(2)} Mins) ${
    curActivity.moving_time > AVG_MTime ? EMOJI_DONE : EMOJI_UNDONE
  }

    AVG HR: ${AVG_HR.toFixed(0)}BPM ${
    curActivity.average_heartrate > AVG_HR ? EMOJI_UP : EMOJI_DOWN
  } 
    AVG Cadence: ${(AVG_CADENCE * 2).toFixed(0)}/Mins ${
    curActivity.average_cadence > AVG_CADENCE ? EMOJI_UP : EMOJI_DOWN
  } 
    AVG Effect Score: ${AVG_SCORE.toFixed(0)} ${
    curActivity.suffer_score > AVG_SCORE ? EMOJI_UP : EMOJI_DOWN
  } 
    `;
}