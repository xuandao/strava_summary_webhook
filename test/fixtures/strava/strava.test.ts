import * as STRAVA from "@/lib/strava";
const detailedActivity = require("./detailedActivity.json")
const summaryActivity = require("./summaryActivity.json")

describe('lib/strava function', () => {
    it('getActivityById', async ()=>{
        // @ts-ignore
        jest.spyOn(STRAVA, 'getActivityById').mockResolvedValue(detailedActivity);
        jest.spyOn(STRAVA, 'getLoggedInAthleteActivities').mockResolvedValue(summaryActivity);

        const curActivity = await STRAVA.getActivityById(10548224944);
        const ActivitiesList = await STRAVA.getLoggedInAthleteActivities('Run', 28);
        const description = STRAVA.generateActivitiesSummary(curActivity, ActivitiesList);

        expect(description).toMatchSnapshot();
    })
});