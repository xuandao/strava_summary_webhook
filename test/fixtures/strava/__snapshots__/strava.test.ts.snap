// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib/strava function getActivityById 1`] = `
"OutDoor SUMMARY:
    Completed: 10 runs
    Total Distance:: 45.93km (Avg.: 4.59km ❌)
    Total Moving Time: 240.77Mins (Avg.: 24.08 Mins) ❌
    AVG HR: 158BPM ⤵️ 
    AVG Cadence: 181/Mins ⤵️ 
    AVG Effect Score: 41 ⤵️

All SUMMARY(IndoorRuns & OutdoorRuns):
    Completed: 23 runs
    Total Distance:: 175.95km (Avg.: 7.65km ❌)
    Total Moving Time: 942.20Mins (Avg.: 40.97 Mins) ❌
    AVG HR: 152BPM ⤵️ 
    AVG Cadence: 178/Mins ⤵️ 
    AVG Effect Score: 44 ⤵️
 power by https://github.com/xuandao/strava_summary_webhook "
`;
